### ISSUE TITLE
Unable to use config values elsewhere

### ISSUE THREAD

**Issue by tyzbit** [kairos-io/kairos#1877]
**Kairos version:**
2.4.0

**CPU architecture, OS, and Version:**
Linux spartacus-e589 6.4.0-4-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.4.13-1 (2023-08-31) x86_64 GNU/Linux

**Describe the bug**
It does not appear to be possible to set the hostname based on the p2p role. When trying to, it results in a hostname of `localhost`.

**To Reproduce**
Create a default config and add:
```yaml
hostname: >-
  {{if eq .Values.p2p.role master }}
    master
  {{ else }}
    worker
  {{ end }}
  -{{ trunc 4 .MachineID }}
```

**Expected behavior**
The masters are given hostnames like `master-e589` and the workers are given hostnames like `worker-e589`.

**Logs**
The logs show a hostname of just `-[MachineID]`

**Additional context**

I've tried `.p2p.role` as well.

My goal is to be able to configure hostnames based on the Kubernetes role (and by extension, Kubernetes nodenames), but I believe in `p2p` mode, the node comes up with its hostname before any Kubernetes setup takes place.

If this is true, then simply being able to specify a nodename for the node as seen in Kubernetes based on the role would be adequate.

**Comment by jimmykarily** [comment]
This page explains what variables yip gives you access to for interpolation: https://github.com/mudler/yip#node-data-interpolation

This is is a feature request (I changed the label).

Is this a request to change the hostname or provide a way to change the node name? Or both?

By the way,  you can see the node's role as a different field, e.g.:

```
~ $ kubectl  get nodes
NAME                                 STATUS   ROLES           AGE   VERSION
kairos-osbuilder-e2e-control-plane   Ready    control-plane   38d   v1.25.3
```

**Comment by tyzbit** [comment]
I think one way I can achieve what I want right now is just to skip out on `p2p.auto` altogether and set either `k3s` or `k3s-agent` sections for each node depending on if it's a worker or control plane node. That shouldn't be too hard to make a Go template to set that according to mac address.

But I think some functionality that would be nice to have to make this easier is to have a stage that specifically runs after `p2p` determines what roles every node has but **before** it runs the actions to install k3s as that role. It should update the config returned by `kairos-agent config get` to retroactively add `p2p.role`. Then I could run a command to change the hostname and ensure the node name ends up what I want as well.

Other users could add other various useful commands at this point, such as mounting a NAS server only if the node is a worker, or registering with a backup server or agent only if the node is a control plane node etc.

**Comment by mudler-agent** [comment]
This now appears to be covered by the provider role hooks. Current `provider-kairos` runs `provider-kairos.bootstrap.before.master` or `provider-kairos.bootstrap.before.worker` after the P2P role is chosen and before the Kubernetes service is configured and started. The stages are also documented in the current cloud-init lifecycle docs. A config can therefore put the hostname or other role-specific commands in the corresponding stage without interpolating `p2p.role`. I suggest closing this as implemented.

**Comment by ci-robbot** [comment]
Triage: the report splits into two asks. The first is still true and unimplemented. The second is already solved, just undocumented, and @mudler's 2024 answer pointed at the wrong stage for it.

### 1. `{{.Values.p2p.role}}` does not work, and still cannot

@tyzbit's template was never going to render, because yip's `.Values` is not the cloud-config. It is only machine facts. Current yip (v1.26.0, `pkg/plugins/common.go`) builds the template context from `sysinfo` plus two extras:

```go
data, err := json.Marshal(&system)     // sysinfo: node, os, memory, ...
...
interpolateOpts["Random"] = utils.RandomString(32)
protectedId, _ := machineid.ProtectedID("yip")
interpolateOpts["ProtectedID"] = protectedId
rendered, err := utils.TemplatedString(s, map[string]interface{}{"Values": interpolateOpts})
```

So `.Values.node.hostname`, `.Values.os.architecture`, `.Values.Random` and `.Values.ProtectedID` resolve; `.Values.p2p.role` resolves to nothing, and `.MachineID` (no `.Values.` prefix) is not a key at all, which is why the log showed a bare `-<machineid>`. Nothing in Kairos feeds the merged cloud-config back into that map. That is the real feature request in this issue and it is still open.

Worth deciding deliberately rather than just implementing: self-referencing config needs an evaluation order (a `hostname` that reads `p2p.role` is fine, a `p2p.role` that reads `hostname` is not), and `.Values` is currently safe to log because it holds no secrets, which stops being true the moment the cloud-config is in there.

### 2. "A stage after p2p decides the role, before it installs k3s as that role" already exists

@mudler suggested `kairos-agent.bootstrap`, but that fires **before** role coordination starts, so the role is still unknown there:

https://github.com/kairos-io/kairos/blob/master/provider/internal/provider/bootstrap.go#L57-L58

The stages that do what @tyzbit asked for are emitted by the role handlers, after the role is elected:

| Stage | Fires |
|---|---|
| `provider-kairos.bootstrap.before.master` | role elected, before the k3s/k0s env file, args, and `svc.Start()` |
| `provider-kairos.bootstrap.before.master/clusterinit` | same, for the cluster-init master |
| `provider-kairos.bootstrap.before.master/ha` | same, for an HA master |
| `provider-kairos.bootstrap.before.worker` | role elected and node token available, before `SetupWorker` and `svc.Start()` |
| `provider-kairos.bootstrap.after.<role>` | after the service is started and enabled |

See `provider/internal/role/p2p/master.go:130` and `provider/internal/role/p2p/worker.go:60`. So the hostname case is:

```yaml
stages:
  provider-kairos.bootstrap.before.master:
    - commands:
        - hostnamectl set-hostname "master-$(cat /etc/machine-id | cut -c1-4)"
  provider-kairos.bootstrap.before.worker:
    - commands:
        - hostnamectl set-hostname "worker-$(cat /etc/machine-id | cut -c1-4)"
```

That also covers the "mount a NAS only on workers" and "register with a backup server only on control planes" examples from the thread.

Caveat on the specific k8s node-name goal: k3s captures the node name when the agent first registers, so renaming the host in `before.<role>` lands before registration and works, while renaming in `after.<role>` does not. Renaming after a node has already joined needs the node deleted and re-registered.

These five stage names are not in the cloud-init reference on kairos.io, which is why this thread went the way it did. Recommend narrowing this issue to ask 1 and opening a docs issue for the role stages, or splitting it. cc @jimmykarily

**Comment by ci-robbot** [comment]
Correction to my last comment: the role stages *are* documented. `docs/architecture/cloud-init.md` lists both of them:

| **Stage** | **Description** |
|---|---|
| `provider-kairos.bootstrap.before.<role>` | The provider fires this stage before starting Kubernetes bootstrap. |
| `provider-kairos.bootstrap.after.<role>`  | The provider fires this stage after Kubernetes bootstrap completes. |

So there is no docs gap for the stages themselves. The only gap is that `<role>` is left unexpanded, and the four values (`master`, `master/clusterinit`, `master/ha`, `worker`) appear nowhere in the docs, which makes the table hard to act on if you do not already know them. That is a one-line docs fix, not a separate issue.

Everything else in my comment stands: ask 2 is solved by these stages, `kairos-agent.bootstrap` is the wrong stage for a role-dependent action because it fires before election, and ask 1 (exposing the cloud-config to yip's `.Values`) is still open and is what this issue should be narrowed to.
